### PR TITLE
ENHANCEMENT More advanced detection of whether outputPath is a parent of the project directory

### PR DESCRIPTION
**Pull request by catbieber**
Revision of: https://github.com/ember-cli/ember-cli/pull/4107

Updated tests so that OS-dependent paths (root, and relative paths) should work on *nix and Windows with the same code.

### CODE REVIEW

**Comment by stefanpenner** [comment]
re-running travis (which timed out do to something quirky)

**Comment by stefanpenner** [comment]
looks like a legit 0.10.x node failure:

```
Build failed.
Object #<Object> has no method 'parse'
TypeError: Object #<Object> has no method 'parse'
    at Class.module.exports.Task.extend.canDeleteOutputPath (/home/travis/build/ember-cli/ember-cli/lib/models/builder.js:47:27)
    at Class.module.exports.Task.extend.clearOutputPath (/home/travis/build/ember-cli/ember-cli/lib/models/builder.js:66:14)
    at Class.module.exports.Task.extend.processBuildResult (/home/travis/build/ember-cli/ember-cli/lib/models/builder.js:95:17)
    at lib$rsvp$$internal$$tryCatch (/home/travis/build/ember-cli/ember-cli/node_modules/rsvp/dist/rsvp.js:489:16)
    at lib$rsvp$$internal$$invokeCallback (/home/travis/build/ember-cli/ember-cli/node_modules/rsvp/dist/rsvp.js:501:17)
    at lib$rsvp$$internal$$publish (/home/travis/build/ember-cli/ember-cli/node_modules/rsvp/dist/rsvp.js:472:11)
    at Object.lib$rsvp$asap$$flush [as _onImmediate] (/home/travis/build/ember-cli/ember-cli/node_modules/rsvp/dist/rsvp.js:1290:9)
    at processImmediate [as _immediateCallback] (timers.js:354:15)
```

File: lib/models/builder.js [hunk 1]
```diff
@@ -36,8 +36,21 @@ module.exports = Task.extend({
...
+  /**
+    Determine whether the output path is safe to delete. If the outputPath
+    appears anywhere in the parents of the project root, the build would 
+    delete the project directory. In this case return `false`, otherwise
+    return `true`.
+  */
   canDeleteOutputPath: function(outputPath) {
-    return this.project.root.indexOf(outputPath) === -1;
+    var rootPathParents = [this.project.root];
+    var parsedPath = path.parse(this.project.root);
```
Comment (stefanpenner) [inline lib/models/builder.js]: `require('path').parse` isn't something available in node 0.10.x

Comment (catbieber) [inline lib/models/builder.js]: Yup, didn't realize that until I saw the failure this morning. Can probably rework using path.dirname() and path.sep instead. Plan to look at that either tonight or tomorrow.

Comment (stefanpenner) [inline lib/models/builder.js]: maybe you can backport the actual parse code into its own module? Might be the safest.

Comment (catbieber) [inline lib/models/builder.js]: I can take a look...I'm really just using dir from the parsed path to make sure I get properly uniform and normalized paths (unfortunately path.normalize() wasn't cutting it, and that parents module that looked cool was also normalizing poorly, in a different way from path.normalize()) so path.dirname() might substitute just fine. If I have trouble with that, I'll look at backporting parse.

**Comment by catbieber** [comment]
@stefanpenner It looks like tests are passing -- let me know if there are more improvements I should make to this patch.

**Comment by stefanpenner** [comment]
@catbieber thanks!  This looks good :)

**Pull request merged**